os/fs/driver/mtd/smart.c

```diff
@@ -2367,6 +2366,37 @@ static inline int smart_getformat(FAR struct smart_struct_s *dev, FAR struct sma
 	return ret;
 }
 
+/****************************************************************************
+ * Name: verify_erased_block
+ *
+ * Description:  Verify all physical sectors in the erased block and adjust
+ *               the freecount of the block if needed due to ERASE FAILURE.
+ *
+ ****************************************************************************/
+
+static void verify_erased_block(FAR struct smart_struct_s *dev, uint16_t block)
+{
+	int ret;
+	int sector;
+	uint32_t readaddr;
+	struct smart_sect_header_s header;
+
+	for (sector = block * dev->sectorsPerBlk; sector < block * dev->sectorsPerBlk + dev->availSectPerBlk; sector++) {
+		readaddr = sector * dev->mtdBlksPerSector * dev->geo.blocksize;
+		ret = MTD_READ(dev->mtd, readaddr, sizeof(struct smart_sect_header_s), (FAR uint8_t *)&header);
+		if (ret != sizeof(struct smart_sect_header_s)) {
+			fdbg("line %d, Error reading phys sector %d\n", __LINE__, sector);
+			return;
+		}
+		if (SECTOR_IS_COMMITTED(header) || SECTOR_IS_RELEASED(header)) {
+			dev->freecount[block]--;
```

Comment:
1. All above operation after MTD_ERASE assuming that erase is finished.
It will change many information, but you are revert that here. 
So, You have to modify another information too.(erasecounts, releasesectors, and so on)
Also if condition should be separated. ( A || B => if (a) else if (b))
2. Here you can check whole of block to verify erase operation totally.

Reply:
Thank you for your comments. Here are my answers.
1. When I inspected another information (freesectors, releasesectors, releasecount),
    freesectors needs only to be updated together since others are already updated right before this function is called.
    And.. I'm sorry but I don't understand why the condition should be seperated.. could you explain in more detail..?
   (Since header should be clean after MTD_ERASE, I regard either one of those two conditions as a sector corruption.)
2. You're also right. But the complete sector verification is also done whenever we find a free physical sector. And as you know, MTD_ERASE is very slow(=expensive) operation so I just simply check headers only here.